Envoy with golang filter randomly crashes

*Title*: Envoy with golang filter randomly crashes

*Description*:
We are using Envoy for quite some time now, I've just added configuration for an envoy filter in a custom built image and now it is randomly crashing.

The configurations for the filter is the one below:

```
- listener_name: frontend
  order: 80
  name: envoy.filters.http.golang
  typed_config:
    "@type": type.googleapis.com/envoy.extensions.filters.http.golang.v3alpha.Config
    library_id: plugin_name
    library_path: /plugin_name.so
    plugin_name: plugin_name
``` 

The plugin is correctly built and executed (I see logs generated by it when a request is received) processing several requests without any issue

*Repro steps*:
Hard to reproduce, the service is up in a testing environment with several user and automated bot using it regularly, at a certain point the services crashes with the stacktrace provided below.

Here is a minimal code 

<details>
  <summary>main.go</summary>

```
const PluginName = "plugin_name"

type EnvConfig struct {
	LogLevel              string `env:"LOG_LEVEL" envDefault:"trace"`
}

func init() {
	http.RegisterHttpFilterConfigFactory(PluginName, ConfigFactory)
}

func ConfigFactory(config interface{}) api.StreamFilterFactory {
	cfg := EnvConfig{}
	if err := env.Parse(&cfg); err != nil {
		panic(fmt.Sprintf("fatal error parsing env: %s", err.Error()))
	}

	log, err := glogger.InitHelper(glogger.InitOptions{Level: cfg.LogLevel})
	if err != nil {
		panic(fmt.Sprintf("fatal error during logger setup: %s", err.Error()))
	}

	log.Trace("envoy golang plugin init completed")
	return func(callbacks api.FilterCallbackHandler) api.StreamFilter {
		return &ConsoleFilter{
			callbacks: callbacks,
			log:       log,
			env:       &cfg,
		}
	}
}

func main() {}
```
</details>
<details>
  <summary>filter.go</summary>

```

type ConsoleFilter struct {
	callbacks api.FilterCallbackHandler
	log       *logrus.Logger
	env       *EnvConfig
}

func (f *ConsoleFilter) sendLocalReplyInternal() api.StatusType {
	headers := make(map[string]string)
	body := ""
	f.callbacks.SendLocalReply(200, body, headers, -1, PluginName)
	return api.LocalReply
}

func (f *ConsoleFilter) DecodeHeaders(header api.RequestHeaderMap, endStream bool) api.StatusType {
	f.log.Trace("decode headers started")
	headerVal, _ := header.Get("myheader")
	f.log.WithFields(logrus.Fields{ "headerVal": headerVal }).Debug("header val from request headers")
	f.log.Trace("decode headers completed")
	return api.Continue
}

func (f *ConsoleFilter) DecodeData(buffer api.BufferInstance, endStream bool) api.StatusType {
	return api.Continue
}

func (f *ConsoleFilter) DecodeTrailers(trailers api.RequestTrailerMap) api.StatusType {
	return api.Continue
}

func (f *ConsoleFilter) EncodeHeaders(header api.ResponseHeaderMap, endStream bool) api.StatusType {
	return api.Continue
}

func (f *ConsoleFilter) EncodeData(buffer api.BufferInstance, endStream bool) api.StatusType {
	return api.Continue
}

func (f *ConsoleFilter) EncodeTrailers(trailers api.ResponseTrailerMap) api.StatusType {
	return api.Continue
}

func (f *ConsoleFilter) OnDestroy(reason api.DestroyReason) {}

```
</details>

*Admin and Stats Output*:

There are a lot of stats, I'm happy to share a subset if needed

*Call Stack*:

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x8 pc=0x7f633749c48c]
goroutine 18 [running, locked to thread]:
github.com/envoyproxy/envoy/contrib/golang/filters/http/source/go/pkg/http.envoyGoFilterOnHttpDestroy(0x0?, 0x0)
	/go/pkg/mod/github.com/envoyproxy/envoy/contrib/golang/filters/http/source/go@v0.0.0-20230112191108-90880d93a9bd/pkg/http/shim.go:169 +0x2c
{"level":"critical","time":"2023-04-04T10:46:49.220619Z","scope":"backtrace","message":"Caught Aborted, suspect faulting address 0x6500000001"}
{"level":"critical","time":"2023-04-04T10:46:49.220691Z","scope":"backtrace","message":"Backtrace (use tools/stack_decode.py to get line numbers):"}
{"level":"critical","time":"2023-04-04T10:46:49.220701Z","scope":"backtrace","message":"Envoy version: e99d61c4596573fbea8b8a9def8b160e138a4018/1.25.3/Clean/RELEASE/BoringSSL"}
{"level":"critical","time":"2023-04-04T10:46:49.221119Z","scope":"backtrace","message":"#0: __restore_rt [0x7f633a95e420]"}
{"level":"critical","time":"2023-04-04T10:46:49.221731Z","scope":"backtrace","message":"#1: runtime.fatalpanic [0x7f633731a14d]"}
{"level":"critical","time":"2023-04-04T10:46:49.222423Z","scope":"backtrace","message":"#2: runtime.gopanic [0x7f633731987b]"}
{"level":"critical","time":"2023-04-04T10:46:49.223094Z","scope":"backtrace","message":"#3: runtime.sigpanic [0x7f633732fcbd]"}
{"level":"critical","time":"2023-04-04T10:46:49.223784Z","scope":"backtrace","message":"#4: github.com/envoyproxy/envoy/contrib/golang/filters/http/source/go/pkg/http.envoyGoFilterOnHttpDestroy [0x7f633749c48c]"}
{"level":"critical","time":"2023-04-04T10:46:49.224425Z","scope":"backtrace","message":"#5: _cgoexp_22941646aa3a_envoyGoFilterOnHttpDestroy [0x7f633749c865]"}
{"level":"critical","time":"2023-04-04T10:46:49.224924Z","scope":"backtrace","message":"#6: runtime.cgocallbackg1 [0x7f63372e9a51]"}
{"level":"critical","time":"2023-04-04T10:46:49.225500Z","scope":"backtrace","message":"#7: runtime.cgocallbackg [0x7f63372e9709]"}
{"level":"critical","time":"2023-04-04T10:46:49.226097Z","scope":"backtrace","message":"#8: runtime.cgocallbackg.abi0 [0x7f633734d171]"}
{"level":"critical","time":"2023-04-04T10:46:49.226575Z","scope":"backtrace","message":"#9: runtime.cgocallback.abi0 [0x7f633734a8b3]"}
ConnectionImpl 0x595fdd38730, connecting_: 0, bind_error_: 0, state(): Closed, read_buffer_limit_: 1048576
socket_: 
  ListenSocketImpl 0x595ff231800, transport_protocol_: raw_buffer
  connection_info_provider_: 
    ConnectionInfoSetterImpl 0x595fefc4838, remote_address_: 10.4.3.106:58766, direct_remote_address_: 10.4.3.106:58766, local_address_: 10.4.1.153:8080, server_name_: 
```

Here are the logs that are printed right befor the crash, `envoy.filters.http.ext_authz` is the filter that is run right before my golang filter, so I started from there right to the crash, after the last log in this batch there is the exact stack trace pasted above. It seems that there's something nasty going on during `golang filter on destroy`, isn't it? 

<details>
<summary>error logs prior the crash</summary>

```
{"level":"trace","time":"2023-04-04T12:48:27.640484Z","scope":"http","message":"[C0][S12414150317462508244] decode headers called: filter=envoy.filters.http.upstream_codec status=0"}
{"level":"trace","time":"2023-04-04T12:48:27.640495Z","scope":"http","message":"[C65][S9326556797505336323] decode headers called: filter=envoy.filters.http.ext_authz status=4"}
{"level":"trace","time":"2023-04-04T12:48:27.640507Z","scope":"main","message":"item added to deferred deletion list (size=4)"}
{"level":"trace","time":"2023-04-04T12:48:27.640512Z","scope":"main","message":"item added to deferred deletion list (size=5)"}
{"level":"debug","time":"2023-04-04T12:48:27.640524Z","scope":"http2","message":"[C3] stream 121 closed: 0"}
{"level":"trace","time":"2023-04-04T12:48:27.640530Z","scope":"main","message":"item added to deferred deletion list (size=6)"}
{"level":"debug","time":"2023-04-04T12:48:27.640534Z","scope":"http2","message":"[C3] Recouping 0 bytes of flow control window for stream 121."}
{"level":"trace","time":"2023-04-04T12:48:27.640543Z","scope":"http2","message":"[C3] dispatched 56 bytes"}
{"level":"trace","time":"2023-04-04T12:48:27.640552Z","scope":"main","message":"clearing deferred deletion list (size=6)"}
{"level":"trace","time":"2023-04-04T12:48:27.640584Z","scope":"connection","message":"[C65] socket event: 3"}
{"level":"trace","time":"2023-04-04T12:48:27.640593Z","scope":"connection","message":"[C65] write ready"}
{"level":"trace","time":"2023-04-04T12:48:27.640599Z","scope":"connection","message":"[C65] read ready. dispatch_buffered_data=0"}
{"level":"trace","time":"2023-04-04T12:48:27.640615Z","scope":"connection","message":"[C65] read returns: 0"}
{"level":"debug","time":"2023-04-04T12:48:27.640621Z","scope":"connection","message":"[C65] remote close"}
{"level":"debug","time":"2023-04-04T12:48:27.640625Z","scope":"connection","message":"[C65] closing socket: 0"}
{"level":"trace","time":"2023-04-04T12:48:27.640712Z","scope":"connection","message":"[C65] raising connection event 0"}
{"level":"debug","time":"2023-04-04T12:48:27.640734Z","scope":"http","message":"[C65][S9326556797505336323] stream reset"}
{"level":"debug","time":"2023-04-04T12:48:27.640921Z","scope":"jwt","message":"Called Filter : onDestroy"}
{"level":"debug","time":"2023-04-04T12:48:27.640943Z","scope":"router","message":"[C0][S12414150317462508244] resetting pool request"}
{"level":"debug","time":"2023-04-04T12:48:27.640950Z","scope":"client","message":"[C4] request reset"}
{"level":"trace","time":"2023-04-04T12:48:27.640955Z","scope":"main","message":"item added to deferred deletion list (size=1)"}
{"level":"debug","time":"2023-04-04T12:48:27.640962Z","scope":"connection","message":"[C4] closing data_to_write=429 type=1"}
{"level":"trace","time":"2023-04-04T12:48:27.641009Z","scope":"connection","message":"[C4] write returns: 429"}
{"level":"debug","time":"2023-04-04T12:48:27.641041Z","scope":"connection","message":"[C4] closing socket: 1"}
{"level":"trace","time":"2023-04-04T12:48:27.641067Z","scope":"connection","message":"[C4] raising connection event 1"}
{"level":"debug","time":"2023-04-04T12:48:27.641073Z","scope":"client","message":"[C4] disconnect. resetting 0 pending requests"}
{"level":"debug","time":"2023-04-04T12:48:27.641080Z","scope":"pool","message":"[C4] client disconnected, failure reason: "}
{"level":"trace","time":"2023-04-04T12:48:27.641088Z","scope":"main","message":"item added to deferred deletion list (size=2)"}
{"level":"trace","time":"2023-04-04T12:48:27.641160Z","scope":"main","message":"item added to deferred deletion list (size=3)"}
{"level":"trace","time":"2023-04-04T12:48:27.641173Z","scope":"main","message":"item added to deferred deletion list (size=4)"}
{"level":"trace","time":"2023-04-04T12:48:27.641177Z","scope":"main","message":"item added to deferred deletion list (size=5)"}
{"level":"trace","time":"2023-04-04T12:48:27.641190Z","scope":"main","message":"item added to deferred deletion list (size=6)"}
{"level":"debug","time":"2023-04-04T12:48:27.641195Z","scope":"http","message":"golang filter on destroy"
```
</details>


@doujiang24 FYI

--- a/contrib/golang/filters/http/source/golang_filter.cc
+++ b/contrib/golang/filters/http/source/golang_filter.cc
@@ -179,6 +179,11 @@ Http::FilterTrailersStatus Filter::encodeTrailers(Http::ResponseTrailerMap& trai
 void Filter::onDestroy() {
   ENVOY_LOG(debug, "golang filter on destroy");
 
+  // do nothing, stream reset may happen before entering this filter.
+  if (req_ == nullptr) {
+    return;
+  }
+
   {
     Thread::LockGuard lock(mutex_);
     if (has_destroyed_) {
@@ -188,7 +193,6 @@ void Filter::onDestroy() {
     has_destroyed_ = true;
   }
 
-  ASSERT(req_ != nullptr);
   auto& state = getProcessorState();
   auto reason = state.isProcessingInGo() ? DestroyReason::Terminate : DestroyReason::Normal;
 

--- a/contrib/golang/filters/http/test/golang_filter_test.cc
+++ b/contrib/golang/filters/http/test/golang_filter_test.cc
@@ -39,9 +39,6 @@ namespace {
 class TestFilter : public Filter {
 public:
   using Filter::Filter;
-  void onDestroy() override {
-    // do nothing
-  }
 };
 
 class GolangHttpFilterTest : public testing::Test {

